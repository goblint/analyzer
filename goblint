#!/bin/bash

DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
python "$DIR/goblint_multishot.py" "$@"
