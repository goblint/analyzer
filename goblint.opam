# This file is generated by dune, edit dune-project instead
opam-version: "2.0"
synopsis: "Static analysis framework for C"
description: """
Goblint is a sound static analysis framework for C programs using abstract interpretation.
It specializes in thread-modular verification of multi-threaded programs, especially regarding data races.
Goblint includes analyses for assertions, overflows, deadlocks, etc and can be extended with new analyses.
"""
maintainer: [
  "Simmo Saan <simmo.saan@gmail.com>"
  "Michael Schwarz <michael.schwarz93@gmail.com>"
  "Karoliine Holter <karoliine.holter@ut.ee>"
]
authors: [
  "Simmo Saan"
  "Michael Schwarz"
  "Julian Erhard"
  "Sarah Tilscher"
  "Karoliine Holter"
  "Ralf Vogler"
  "Kalmer Apinis"
  "Vesal Vojdani"
]
license: "MIT"
tags: [
  "program analysis"
  "program verification"
  "static analysis"
  "abstract interpretation"
  "C"
  "data race analysis"
  "concurrency"
]
homepage: "https://goblint.in.tum.de"
doc: "https://goblint.readthedocs.io/en/latest/"
bug-reports: "https://github.com/goblint/analyzer/issues"
depends: [
  "dune" {>= "3.13"}
  "ocaml" {>= "4.14"}
  "goblint-cil" {>= "2.0.5"}
  "batteries" {>= "3.5.1"}
  "zarith" {>= "1.10"}
  "yojson" {>= "2.0.0"}
  "qcheck-core" {>= "0.19"}
  "ppx_deriving" {>= "6.0.2"}
  "ppx_deriving_hash" {>= "0.1.2"}
  "ppx_deriving_yojson" {>= "3.7.0"}
  "ppx_blob" {>= "0.8.0"}
  "ppxlib" {>= "0.30.0"}
  "ounit2" {with-test}
  "qcheck-ounit" {with-test}
  "odoc" {with-doc}
  "fpath"
  "dune-site"
  "dune-build-info"
  "json-data-encoding"
  "jsonrpc" {>= "1.12"}
  "sha" {>= "1.12"}
  "fileutils" {>= "0.6.4"}
  "cpu"
  "arg-complete" {>= "0.2.1"}
  "yaml" {>= "3.0.0"}
  "uuidm"
  "catapult"
  "catapult-file"
  "conf-gmp" {>= "3"}
  "conf-ruby" {with-test}
  "benchmark" {with-test}
  "conf-gcc"
]
depopts: [
  "apron" {>= "v0.9.15"}
  "z3"
]
conflicts: [
  "result" {< "1.5"}
  "ez-conf-lib" {= "1"}
]
build: [
  ["dune" "subst"] {dev}
  [
    "dune"
    "build"
    "-p"
    name
    "-j"
    jobs
    "--promote-install-files=false"
    "@install"
    "@runtest" {with-test}
    "@doc" {with-doc}
  ]
  ["dune" "install" "-p" name "--create-install-files" name]
]
dev-repo: "git+https://github.com/goblint/analyzer.git"
# on `dune build` goblint.opam will be generated from goblint.opam.template and dune-project
# also remember to generate/adjust goblint.opam.locked!
available: os-family != "bsd" & os-distribution != "alpine" & (arch != "arm64" | os = "macos")
pin-depends: [
  # published goblint-cil 2.0.5 is currently up-to-date, but pinned for reproducibility
  [ "goblint-cil.2.0.5" "git+https://github.com/goblint/cil.git#f5ee39bd344dc74e2a10e407d877e0ddf73c9c6f" ]
  # pinned for stability (https://github.com/goblint/analyzer/issues/1520), remove after new camlidl release
  [ "camlidl.1.12" "git+https://github.com/xavierleroy/camlidl.git#1c1e87e3f56c2c6b3226dd0af3510ef414b462d0" ]
  # pinned for stability (https://github.com/goblint/analyzer/issues/1520), remove after new apron release
  [ "apron.v0.9.15" "git+https://github.com/antoinemine/apron.git#418a217c7a70dae3f422678f3aaba38ae374d91a" ]
]
depexts: [
  ["libgraph-easy-perl"] {os-distribution = "ubuntu" & with-test}
]
post-messages: [
  "Do not benchmark Goblint on OCaml 5 (https://goblint.readthedocs.io/en/latest/user-guide/benchmarking/)." {ocaml:version >= "5.0.0"}
]
x-ci-accept-failures: [
  "macos-homebrew" # newer MacOS headers cannot be parsed (https://github.com/ocaml/opam-repository/pull/26307#issuecomment-2258080206)
  "opensuse-tumbleweed" # not GNU diff, so some cram tests fail (https://discuss.ocaml.org/t/opensuse-and-opam-tests/14641/2)
]
