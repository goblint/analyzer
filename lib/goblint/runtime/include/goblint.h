void __goblint_check(_Bool exp);
void __goblint_assume(_Bool exp);
void __goblint_assert(_Bool exp);

void __goblint_assume_join(/* pthread_t thread */); // undeclared argument to avoid pthread.h interfering with Linux kernel headers
void __goblint_globalize(void *ptr);

void __goblint_split_begin(int exp);
void __goblint_split_end(int exp);

void __goblint_bounded(unsigned long long exp);