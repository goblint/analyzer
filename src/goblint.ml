open Goblint_lib
open GobConfig
open Maingoblint

(** the main function *)
let main () =
  try
    Maingoblint.parse_arguments ();
    Cilfacade.init ();

    (* Timing. *)
    Maingoblint.reset_stats ();
    if get_bool "dbg.timing.enabled" then (
      let tef_filename = get_string "dbg.timing.tef" in
      if tef_filename <> "" then
        Goblint_timing.setup_tef tef_filename;
      Timing.Default.start {
        cputime = true;
        walltime = true;
        allocated = true;
        count = true;
        tef = true;
      };
      Timing.Program.start {
        cputime = false;
        walltime = false;
        allocated = false;
        count = false;
        tef = true;
      }
    );

    handle_extraspecials ();
    GoblintDir.init ();

    Logs.debug "%s" (GobUnix.localtime ());
    Logs.debug "%s" GobSys.command_line;
    (* When analyzing a termination specification, activate the termination analysis before pre-processing. *)
    if get_string "ana.specification" <> "" then AutoSoundConfig.enableAnalysesForTerminationSpecification ();
    if AutoTune.isActivated "termination" then AutoTune.focusOnTermination ();
    let file = lazy (Fun.protect ~finally:GoblintDir.finalize preprocess_parse_merge) in
    if get_bool "server.enabled" then (
      let file =
        if get_bool "server.reparse" then
          None
        else
          Some (Lazy.force file)
      in
      Server.start file
    )
    else (
      let file = Lazy.force file in
      let changeInfo =
        if GobConfig.get_bool "incremental.load" || GobConfig.get_bool "incremental.save" then
          diff_and_rename file
        else
          None
      in
      (* This is run independant of the autotuner being enabled or not to be sound for programs with longjmp *)
      AutoSoundConfig.activateLongjmpAnalysesWhenRequired ();
      if get_string "ana.specification" <> "" then AutoSoundConfig.enableAnalysesForSpecification ();
      if get_bool "ana.autotune.enabled" then AutoTune.chooseConfig file;
      file |> do_analyze changeInfo;
      do_html_output ();
      do_gobview file;
      do_stats ();
      Goblint_timing.teardown_tef ();
      (* TODO: generalize exit codes for AnalysisState.unsound_both_branches_dead? *)
      if !AnalysisState.verified = Some false then exit 3 (* verifier failed! *)
    )
  with
  | Stdlib.Exit ->
    do_stats ();
    Goblint_timing.teardown_tef ();
    exit 1
  | Sys.Break -> (* raised on Ctrl-C if `Sys.catch_break true` *)
    do_stats ();
    Printexc.print_backtrace stderr;
    Logs.error "%s" (MessageUtil.colorize ~fd:Unix.stderr ("{RED}Analysis was aborted by SIGINT (Ctrl-C)!"));
    Goblint_timing.teardown_tef ();
    exit 131 (* same exit code as without `Sys.catch_break true`, otherwise 0 *)
  | Timeout.Timeout ->
    do_stats ();
    Logs.error "%s" (MessageUtil.colorize ~fd:Unix.stderr ("{RED}Analysis was aborted because it reached the set timeout of " ^ get_string "dbg.timeout" ^ " or was signalled SIGPROF!"));
    Goblint_timing.teardown_tef ();
    exit 124
  | Svcomp.Error msg ->
    do_stats ();
    Witness.print_svcomp_result ("ERROR (" ^ msg ^ ")");
    Goblint_timing.teardown_tef ();
    exit 1

(* We do this since the evaluation order of top-level bindings is not defined, but we want `main` to run after all the other side-effects (e.g. registering analyses/solvers) have happened. *)
let () = at_exit main
