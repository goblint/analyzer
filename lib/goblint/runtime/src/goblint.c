#include "goblint.h"
#include <pthread.h>

// Empty implementations (instead of asserts) because annotating documentation promises no-op right now.

void __goblint_check(_Bool exp) {

}

void __goblint_assume(_Bool exp) {

}

void __goblint_assert(_Bool exp) {

}


void __goblint_assume_join(pthread_t thread) {

}


void __goblint_split_begin(int exp) {

}

void __goblint_split_end(int exp) {

}

void __goblint_bounded(unsigned long long exp) {

}