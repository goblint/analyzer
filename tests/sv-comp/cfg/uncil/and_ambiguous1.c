#include <goblint.h>

int main() {
    int a, b;

    __goblint_split_begin(a);
    if (a && b) {
        return 1;
    }
    else {
        return 0;
    }
}
