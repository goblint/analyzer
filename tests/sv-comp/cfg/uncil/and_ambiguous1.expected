                             ┌──────────────────────────────────┐
                             │                _                 │
                             └──────────────────────────────────┘
                               │
                               │ Entry main
                               ▼
                             ┌──────────────────────────────────┐
                             │                _                 │
                             └──────────────────────────────────┘
                               │
                               │ Proc '__goblint_split_begin(a)'
                               ▼
┌───┐  Test (a && b,false)   ┌──────────────────────────────────┐  Test (a && b,false)   ┌───┐
│ _ │ ◀───────────────────── │                _                 │ ─────────────────────▶ │ _ │
└───┘                        └──────────────────────────────────┘                        └───┘
  │                            │                                                           │
  │                            │ Test (a && b,true)                                        │
  │                            ▼                                                           │
  │                          ┌──────────────────────────────────┐                          │
  │                          │                _                 │                          │
  │                          └──────────────────────────────────┘                          │
  │                            │                                                           │
  │                            │ Ret (Some 1, main)                                        │
  │                            ▼                                                           │
  │   Ret (Some 0, main)     ┌──────────────────────────────────┐  Ret (Some 0, main)      │
  └────────────────────────▶ │                _                 │ ◀────────────────────────┘
                             └──────────────────────────────────┘
